File: src/problemset/linkedList/ReverseLinkedList.java
package problemset.linkedList;

/**
 * @date: 2020-03-31
 * @time: 15:05:05
 * @author: MoeexT
 * @problem: 206
 * @url: https://leetcode-cn.com/problems/reverse-linked-list/solution/fan-zhuan-lian-biao-by-leetcode/
 */

import util.ListNode;

public class ReverseLinkedList {
    /*
     * @time: 0ms 100.00%
     * @mem: 40.1MB 5.13%
     */
    /*
    public ListNode reverseList(ListNode head) {
        ListNode pre = null, cur = head;
        while (head != null) {
            cur = head;
            head = head.next;
            cur.next = pre;
            pre = cur;
        }
        return cur;
    }*/

    /*
     *
     * @time: 0ms 100.00%
     * @mem: 38.1MB 81.05%
     */
    public ListNode reverseList(ListNode head) {
        return reverse(null, head);
    }

    private ListNode reverse(ListNode pre, ListNode cur) {
        if (cur == null) return pre;
        ListNode nxt = cur.next;
        cur.next = pre;
        return reverse(cur, nxt);
    }

    public static void main(String[] args) {
        ReverseLinkedList obj = new ReverseLinkedList();
        ListNode node = ListNode.generateList(new int[]{1, 2, 3, 4, 5});
        ListNode.printList(node);
        ListNode.printList(obj.reverseList(node));
    }
}

File: src/problemset/linkedList/ReverseLinkedList.py
"""
problem 206
https://leetcode-cn.com/problems/reverse-linked-list/
反转一个单链表
"""

from utils.link_list import ListNode, LinkList, show_list


class Solution:
    # def reverseList(self, head: ListNode) -> ListNode:
    #     if head is None:
    #         return None
    #
    #     stack = []
    #     while head:
    #         stack.insert(0, head)
    #         head = head.next
    #
    #     res = stack[0]
    #     cursor = res
    #     for node in stack[1:]:
    #         cursor.next = node
    #         cursor = cursor.next
    #     cursor.next = None
    #
    #     return res
    def reverseList(self, head: ListNode) -> ListNode:
        p, rev = head, None
        while p:
            print("rev:", rev)
            print("p:", p)
            rev, rev.next, p = p, rev, p.next
            print("rev:", rev)
            print("p:", p)
            print('-----------------------')
        return rev


if __name__ == '__main__':
    list_head = LinkList.single_list(5)
    sol = Solution()
    res = sol.reverseList(list_head)


